network-mux/src/Network/Mux.hs

```diff
@@ -157,26 +208,68 @@ muxStart tracer peerid app bearer = do
                               cnt
 
         return $ case app of
-          MuxInitiatorApplication initiator ->
-            [ ( initiator peerid mpdId initiatorChannel >> mpsJobExit cnt
+          MuxInitiatorApplication _ _ ->
+            [ ( mpsInitiatorBrooder itbl mpdId app initiatorChannel
               , show mpdId ++ " Initiator" )]
-          MuxResponderApplication responder ->
-            [ ( responder peerid mpdId responderChannel >> mpsJobExit cnt
-              , show mpdId ++ " Responder" )]
-          MuxInitiatorAndResponderApplication initiator responder ->
-            [ ( initiator peerid mpdId initiatorChannel >> mpsJobExit cnt
+          MuxResponderApplication _ _ ->
+            [ (mpsResponderBrooder pmss rtbl mpdId app responderChannel
+            , show mpdId ++ " Responder" )]
+          MuxInitiatorAndResponderApplication _ _ _ _ ->
+            [ ( mpsInitiatorBrooder itbl mpdId app initiatorChannel
               , show mpdId ++ " Initiator" )
-            , (responder peerid mpdId responderChannel >> mpsJobExit cnt
+            , (mpsResponderBrooder pmss rtbl mpdId app responderChannel
               , show mpdId ++ " Responder" )
             ]
 
-    -- cnt represent the number of SDUs that are queued but not yet sent.  Job
-    -- threads will be prevented from exiting until all SDUs have been
-    -- transmitted unless an exception/error is encounter. In that case all
-    -- jobs will be cancelled directly.
-    mpsJobExit :: StrictTVar m Int -> m ()
-    mpsJobExit cnt = do
+    -- Start and restart responder side miniprotocols on demand.
+    mpsResponderBrooder
+      :: HasResponder appType ~ True
+      => PerMuxSharedState ptcl m
+      -> MiniProtocolResponderControlTable ptcl m b
+      -> ptcl
+      -> MuxApplication appType peerid ptcl m a b
+      -> Channel m
+      -> m ()
```

Comment:
You could return `Void` instead of `()`.  We use it in other places where `forever` or other infinite loops are used.

Reply:
It needs to match the type of other mux jobs, some of which can return.